scapy/layers/usb.py

```diff
@@ -0,0 +1,166 @@
+        return p + pay
+
+
+class USBpcapTransferIsochronous(Packet):
+    name = "USBpcap Transfer Isochronous"
+    fields_desc = [
+        LEIntField("offset", 0),
+        LEIntField("length", 0),
+        LEIntEnumField("usbd_status", 0x0, _usbd_status_codes)
+    ]
+
+
+class USBpcapTransferInterrupt(Packet):
+    name = "USBpcap Transfer Interrupt"
+    fields_desc = [
+        LEIntField("startFrame", 0),
+        LEIntField("numberOfPackets", 0),
+        LEIntField("errorCount", 0)
+    ]
+
+
+class USBpcapTransferControl(Packet):
+    name = "USBpcap Transfer Control"
+    fields_desc = [
+        ByteField("stage", 0)
+    ]
+
+
+bind_layers(USBpcap, USBpcapTransferIsochronous, transfer=0)
+bind_layers(USBpcap, USBpcapTransferInterrupt, transfer=1)
+bind_layers(USBpcap, USBpcapTransferControl, transfer=2)
+
+conf.l2types.register(DLT_USBPCAP, USBpcap)
+
+
+def _extcap_call(prog, args, keyword, values):
+    """Function used to call a program using the extcap format,
+    then parse the results"""
+    data = sp.Popen(
+        [prog] + args,
+        stdout=sp.PIPE, stderr=sp.STDOUT
+    ).communicate()[0]
+    print(data)
+    res = []
+    for ifa in data.split("\n"):
+        ifa = ifa.strip()
+        if not ifa.startswith(keyword):
+            continue
+        res.append(tuple([re.search(r"{%s=([^}]*)}" % val, ifa).group(1)
+                   for val in values]))
+    return res
+
+
+if WINDOWS:
+    def _usbpcap_check():
+        if not conf.prog.usbpcapcmd:
+            raise OSError("USBpcap is not installed ! (USBpcapCMD not found)")
+    def get_usbpcap_interfaces():
+        """Return a list of available USBpcap interfaces"""
+        _usbpcap_check()
+        return _extcap_call(
+            conf.prog.usbpcapcmd,
+            ["--extcap-interfaces"],
+            "interface",
+            ["value", "display"]
+        )
+
+    def get_usbpcap_devices(iface, enabled=True):
+        """Return a list of devices on an USBpcap interface"""
+        _usbpcap_check()
+        devices = _extcap_call(
+            conf.prog.usbpcapcmd,
+            ["--extcap-interface",
+             iface,
+             "--extcap-config"],
+            "value",
+            ["value", "display", "enabled"]
+        )
+        devices = [(dev[0],
+                    dev[1],
+                    dev[2] == "true") for dev in devices]
+        if enabled:
+            return [dev for dev in devices if dev[2]]
+        return devices
+
+    class USBpcapSocket(SuperSocket):
+        desc = "read packets at layer 2 using tcpdump"
+
+        def __init__(self, iface=None, *args, **karg):
+            _usbpcap_check()
+            if iface is None:
+                warning("Available interfaces: [%s]" %
+                        " ".join(x[0] for x in get_usbpcap_interfaces()))
+                raise NameError("No interface specified !")
+            self.outs = None
+            args = ['-d', iface, '-b', str(MTU), '-A', '-o', '-']
```

Comment:
What value the MTU typically is? The -b option (--bufferlen) in USBPcapCMD determines the internal kernel mode buffer. If you set it too low, you'll most likely experience that not all packets are captured (and worse, it's not really indicated). This should be suitably high, however due to the fact that it must be allocated from NonPagedPool, it cannot be too big.

Reply:
This (64 KiB - 1) is smaller than the default bufferlen of 1 MiB. With such buffer size it's likely that you'll completely miss some IRPs, as according to https://github.com/desowin/usbpcap/issues/26 there are packets larger than 65535 bytes in the wild.

Reply:
Thanks for the tip, I've updated it to the max value

Reply:
It really depends on what computers it will be run. Check the https://blogs.technet.microsoft.com/markrussinovich/2009/03/10/pushing-the-limits-of-windows-paged-and-nonpaged-pool/ for table listing the NonPagedPool size. The maximum that USBPcap allows (128 MiB) is sometimes more than the total amount of NonPagedPool available!

Reply:
Wow sounds complicated :/ Thanks for the tip ! I think I’ll just leave it as so, I don’t think many people still use Windows XP nor 32-bits systems